packages/itwinui-react/src/core/Tabs/Tabs.tsx

```diff
@@ -526,85 +511,360 @@ export const Tabs = (props: TabsProps) => {
+// Tabs.TabDescription component
+
+const TabsTabDescription = polymorphic('iui-tab-description');
+TabsTabDescription.displayName = 'Tabs.TabDescription';
+
+// ----------------------------------------------------------------------------
+// Tabs.Actions component
+
+type TabsActionsOwnProps = {}; // eslint-disable-line @typescript-eslint/ban-types
+
+const TabsActions = React.forwardRef((props, ref) => {
+  const { className, children, ...rest } = props;
+
+  const { type } = useSafeContext(TabsContext);
+
+  if (type !== 'pill') {
+    return (
       <Box
-        as='ul'
-        className={cx(
-          'iui-tabs',
-          `iui-${type}`,
-          {
-            'iui-green': color === 'green',
-            'iui-animated': type !== 'default' && isClient,
-            'iui-not-animated': type !== 'default' && !isClient,
-            'iui-large': hasSublabel,
-          },
-          tabsClassName,
-        )}
-        data-iui-overflow={overflowOptions?.useOverflow}
-        data-iui-scroll-placement={scrollingPlacement}
-        role='tablist'
-        ref={refs}
-        onKeyDown={onKeyDown}
+        className={cx('iui-tabs-actions-wrapper', className)}
+        ref={ref}
         {...rest}
       >
-        {labels.map((label, index) => createTab(label, index))}
+        {children}
       </Box>
+    );
+  } else {
+    return <></>;
+  }
+}) as PolymorphicForwardRefComponent<'div', TabsActionsOwnProps>;
+TabsActions.displayName = 'Tabs.Actions';
 
-      {actions && (
-        <Box className='iui-tabs-actions-wrapper'>
-          <Box className='iui-tabs-actions'>{actions}</Box>
-        </Box>
-      )}
+// ----------------------------------------------------------------------------
+// Tabs.Action component
 
-      {children && (
-        <Box
-          className={cx('iui-tabs-content', contentClassName)}
-          role='tabpanel'
-        >
-          {children}
-        </Box>
-      )}
-    </Box>
+type TabsActionOwnProps = {}; // eslint-disable-line @typescript-eslint/ban-types
+
+const TabsAction = React.forwardRef((props, ref) => {
+  const { className, children, ...rest } = props;
+
+  const { type } = useSafeContext(TabsContext);
+
+  if (type !== 'pill') {
+    return (
+      <Box className={cx('iui-tabs-actions', className)} ref={ref} {...rest}>
+        {children}
+      </Box>
+    );
+  } else {
+    return <></>;
+  }
+}) as PolymorphicForwardRefComponent<'div', TabsActionOwnProps>;
+TabsAction.displayName = 'Tabs.Action';
+
+// ----------------------------------------------------------------------------
+// Tabs.Panels component
+
+const TabsPanels = ({
+  children,
+}: {
+  children: React.ReactNode[] | React.ReactNode;
+}) => {
+  const { currentActiveIndex } = useSafeContext(TabsContext);
+
+  const items = Array.isArray(children) ? children : [children];
+
+  return (
+    <>
+      {items.map((item, index) => {
+        return (
+          <TabsContext.Provider
```

Comment:
Why do you have two TabsContext.Providers?

Reply:
The context provider for the wrapper is the main one that passes different variables to `Tabs.TabList`.

The context provider for the tabs list passes an updated `currentActiveIndex`, an `onClick` function that was created in `Tabs.TabList`, a `focusedIndex` variable created in `Tabs.TabList`, and an `index` variable to each of the `Tabs.Tab` children so that each child can be indexed properly. The same thing happens when passing `index` from `Tabs.Panels` to `Tabs.Panel`.

Is there any issues I'm not aware of that occur when nesting context providers this way?

Reply:
Try to use ThemeProvider only once.
You are just looping the value in Panel:
![image](https://github.com/iTwin/iTwinUI/assets/83585998/c7e6cef6-f93a-4d60-aebd-9e33ff7054eb)

Reply:
`Panel` only gets the `currentActiveIndex` value from its grandparent (`Tabs`) if the parent (`Panels`) doesn't use a context provider. Since `Panels` needs to use a context provider to pass `index` to `Panel`, the `currentActiveIndex` must be passed through this context as well.

Can you be more specific about your ThemeProvider comment?

Reply:
i believe Greta meant TabsContext.Provider, not ThemeProvider

---

here are my general thoughts:
- it doesn't make sense to override context value completely. If the two instances of the context serve different purposes, then you probably need two different contexts
- it's probably not a good idea to create several different contexts inside a loop when most of the values are shared, which brings me to my next point:
- We might want to move a lot of the code and props from `TabsComponent` to `Tabs.Tab`. Each tab knows about itself, so less code will be necessary to make it do similar work. For example, the `activeIndex` prop could be removed from the top level component and instead passed to the individual tabs.
  ```jsx
  <Tabs>
    <Tabs.Tab>...</Tabs.Tab>
    <Tabs.Tab isActive>...</Tabs.Tab>
    <Tabs.Tab>...</Tabs.Tab>
  </Tabs>
  ```

Reply:
I get why multiple contexts should be used for this component, and I agree that all of the props from `Tabs` aside from `orientation` can at least be passed into `Tabs.TabList` from the user. I'll try your approach with using `isActive` in `Tabs.Tab` to get the active tab.

Reply:
After looking into this further, I'm not sure if this is a better approach for determining which tab is active. Would each tab have it's own `onActiveChange` prop? This method also allows users to have more than one active tab at a time. Also, `Tabs.TabList` still needs to know which tab is active relative to the other tabs for keyboard handling, so I think it's best if `activeIndex` and `onTabSelected` are props for `Tabs.TabList`.

Reply:
> Would each tab have it's own `onActiveChange` prop?

yes

> This method also allows users to have more than one active tab at a time

user-error 🤷‍♂️ i'd expect users to know how to correctly maintain state, especially when we'll have correct examples.

> Also, `Tabs.TabList` still needs to know which tab is active relative to the other tabs for keyboard handling

that code can also be moved into `Tabs.Tab`, and would likely be better that way

Reply:
> i believe Greta meant TabsContext.Provider, not ThemeProvider

Oops.

Reply:
- `TabsContext` has been broken up into `TabsTabListContext` and `TabsPanelsContext` so that each context only uses `.Provider` once
- `onTabSelected` and `activeIndex` props have been removed
- `isActive` and `onActiveChange` props have been added to `Tabs.Tab`
- All keyboard navigation is now handled through `Tabs.Tab`